Client - Get Outputs Id Refactor

rebinmannady93 opened the pull request:
# Description of change

Fixes #276 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

test_get_outputs_id.c

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

rebinmannady93 committed:
Outputs refactor.

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
Removing unwanted comments.

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
1. ADDRESS_NFT_HEX_BYTES -> BECH32_ENCODED_NFT_ADDRESS
2. ADDRESS_ALIAS_HEX_BYTES -> BECH32_ENCODED_ALIAS_ADDRESS
3. ADDRESS_FOUNDRY_HEX_BYTES -> BECH32_ENCODED_ALIAS_ADDRESS

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
1. Added macros TO_HEX_BYTES(x) and TO_STR_BYTES(x).
2. ADDRESS_NFT_HEX_BYTES -> TO_HEX_BYTES(NFT_ID_BYTES)
3. ADDRESS_ALIAS_HEX_BYTES -> TO_HEX_BYTES(ALIAS_ID_BYTES)
4. ADDRESS_FOUNDRY_HEX_BYTES) -> TO_HEX_BYTES(FOUNDRY_ID_BYTES)

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
Query parameter structure.

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
Added src/core/utils/macros.h

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
Some comments updated.

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

matjazv commented on src/client/api/restful/get_outputs_id.c:
```diff
@@ -96,23 +101,32 @@ int deser_outputs(char const *const j_str, res_outputs_id_t *res) {
+  cJSON *json_cursor = cJSON_GetObjectItemCaseSensitive(json_obj, JSON_KEY_CURSOR);
+  if (json_cursor != NULL) {
+    if (cJSON_IsString(json_cursor) && (json_cursor->valuestring != NULL)) {
+      res->u.output_ids->cursor = malloc(strlen(json_cursor->valuestring) + 1);
+      memset(res->u.output_ids->cursor, 0, strlen(json_cursor->valuestring) + 1);
```
```suggestion
      res->u.output_ids->cursor = calloc(1, strlen(json_cursor->valuestring) + 1);
```
Instead of calling two functions, a `calloc` function may be used which allocates memory and initializes it with 0.

rebinmannady93 replied:
Used memset as we have not used calloc anywhere in our lib. Changed to calloc, better.

matjazv commented on src/client/api/restful/get_outputs_id.c:
```diff
@@ -258,14 +271,14 @@ int get_outputs_from_nft_id(iota_client_conf_t const *conf, char const nft_id[],
   // compose restful api command
-  char cmd_buffer[67] = {0};  // 67 = max size of api path(28) + ADDRESS_NFT_HEX_BYTES(40) + 1
-  int snprintf_ret = snprintf(cmd_buffer, sizeof(cmd_buffer), "/api/plugins/indexer/v1/nft/%s", nft_id);
+  char cmd_buffer[70] = {0};  // 70 = max size of api path(29) + BIN_TO_HEX_BYTES(NFT_ID_BYTES)(40) + 1
```
What `+ 1` represents in a comment? NULL string terminator? If so, then `BIN_TO_HEX_STR_BYTES(NFT_ID_BYTES)` could be used.

rebinmannady93 replied:
For (id_len != BIN_TO_HEX_BYTES(ALIAS_ID_BYTES)), we cannot use BIN_TO_HEX_STR_BYTES(NFT_ID_BYTES). + 1 is for null character for the cmd_buffer.

matjazv replied:
Yes this is true.
I just mean to change comment like this:
```suggestion
  char cmd_buffer[70] = {0};  // 70 = max size of api path(29) + BIN_TO_HEX_STR_BYTES(NFT_ID_BYTES)(41)
```

rebinmannady93 replied:
But actually, the null character is not for BIN_TO_HEX_STR_BYTES(NFT_ID_BYTES). It is for the cmd_buffer to have null character.

matjazv replied:
Ok. Then just ignore my suggestion.

matjazv commented on src/core/utils/macros.h:
```diff
@@ -0,0 +1,14 @@
+// Get the string bytes of the given binary
+#define BIN_TO_STR_BYTES(x) (x + 1)
+
+#endif
```
No new line at the end of a file.

rebinmannady93 replied:
Done.

matjazv commented on tests/client/api_restful/test_outputs_id.c:
```diff
@@ -15,31 +15,27 @@ void tearDown(void) {}
+  TEST_ASSERT(res->u.output_ids->page_size == 2);
   TEST_ASSERT(res_outputs_output_id_count(res) == 2);
   TEST_ASSERT(utarray_len(res->u.output_ids->outputs) == 2);
   TEST_ASSERT_EQUAL_MEMORY("1c6943b0487c92fd057d4d22ad844cc37ee27fe6fbe88e5ff0d20b2233f75b9d0005",
```
This validates data without checking for NULL string terminator. Because message id is a string, it is more appropriate to use `TEST_ASSERT_EQUAL_STRING`.

rebinmannady93 replied:
Done.

rebinmannady93 committed:
1. Using calloc instead of malloc and memset.
2. New line and macros.h EOF.
3. Using TEST_ASSERT_EQUAL_STRING instead of TEST_ASSERT_EQUAL_MEMORY for checking string equality.

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

oopsmonk commented on src/client/api/restful/get_outputs_id.c:
```diff
@@ -96,23 +101,31 @@ int deser_outputs(char const *const j_str, res_outputs_id_t *res) {
+  if (json_cursor != NULL) {
+    if (cJSON_IsString(json_cursor) && (json_cursor->valuestring != NULL)) {
+      res->u.output_ids->cursor = calloc(strlen(json_cursor->valuestring) + 1, 1);
+      strncpy(res->u.output_ids->cursor, json_cursor->valuestring, strlen(json_cursor->valuestring));
```
refer to disscution here: https://github.com/iotaledger/iota.c/pull/286#discussion_r803360717
`memset` or `calloc` is not necessary, the buffer is immediately used and padded with null terminator via `strncpy` in next line. I'd suggest using `malloc`, it's faster and zeroing out is not needed in this case.

rebinmannady93 replied:
Done.

oopsmonk commented on src/client/api/restful/get_outputs_id.h:
```diff
@@ -35,6 +34,35 @@ typedef struct {
+ *
+ */
+typedef struct {
+  char *address;                          ///< The Bech32-encoded address that should be searched for
```
the max length of addresses are fixed, it can be reserved for reducing dymanic allocation. like `address`, `dust_ret_address`, `expiration_return_address`, `bech32_sender`, `tag`  in the structure.
and also rearange the items order to make it packed as posiable.
personally, `expiration_return_address` looks to long, maybe rename to `expir_ret_address`?

rebinmannady93 replied:
Accidently commited the outputs_query_params_t struct. It is development under progress. Removed it. Will take care of the comments in the next PR.

matjazv replied:
Actually, it would be better if we would have a list for all query parameters like we have it for unlock conditions.
Each query parameter would have its own structure with a type inside.

If for example a user would just like to add address as a query parameter, to use whole `outputs_query_params_t` structure would take too much memory space in a microcontroller.

rebinmannady93 replied:
Okay Noted.

rebinmannady93 committed:
1. Used malloc instead of calloc.
2. Copying including NULL character using strncpy.

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
Removed accidently added querry params.

Signed-off-by: rebinmannady93 <rabin.muhammed@iota.org>

rebinmannady93 committed:
Merge branch 'tokenization-sc' into outputs_refactor
